File: wilcoxon.py
#!/usr/bin/env python3

# wilcoxon.py
# Mike Zhang
# 02-03-2019

import pandas as pd
import re

def changes(both, src , tra, lang, year):

    print('LANGUAGE: ', lang)
    both = both.sort_values(by=['Z.BOTH'], ascending=False)
    src = src.sort_values(by=['Z.SOURCE'], ascending=False)
    trans = tra.sort_values(by=['Z.TRANS'], ascending=False)
    both_copy = pd.DataFrame(data=None, index=both['SYS'], columns=both['SYS'])
    src_copy = pd.DataFrame(data=None, index=src['SYS'], columns=src['SYS'])
    tra_copy = pd.DataFrame(data=None, index=trans['SYS'], columns=trans['SYS'])

    print()
    print('WMT INPUT')
    with open('original_input.txt') as f:
        for line in re.findall(year+'_'+lang+'.csv(.*?)wilcoxon',f.read(), re.S):
            line = line.replace('[1]','').replace('"', '').strip().split('\n')
            it = iter(line)
            for systems in it:
                sys = systems.split()[1:]
                value = next(it).split()[-1]
                if '*' not in value:
                    value = float(value)
                both_copy.at[sys[0],sys[1]] = value
                both_copy = both_copy.fillna('-')
            if 'HUMAN' in both_copy:
                both_copy = both_copy.drop(index='HUMAN', columns='HUMAN')

        print(both_copy.round(5))

    f.close()

    print()
    print('ONLY ORIGINAL INPUT')
    with open('source_input.txt') as f:
        for line in re.findall(year+'_'+lang+'.csv(.*?)wilcoxon',f.read(), re.S):
            line = line.replace('[1]','').replace('"', '').strip().split('\n')
            it = iter(line)
            for systems in it:
                sys = systems.split()[1:]
                value = next(it).split()[-1]
                if '*' not in value:
                    value = float(value)
                src_copy.at[sys[0],sys[1]] = value
                src_copy = src_copy.fillna('-')
            if 'HUMAN' in src_copy:
                src_copy = src_copy.drop(index='HUMAN', columns='HUMAN')

        print(src_copy.round(5))

    f.close()

    print()
    print('ONLY TRANSLATIONESE INPUT')
    with open('trans_input.txt') as f:
        for line in re.findall(year+'_'+lang+'.csv(.*?)wilcoxon',f.read(), re.S):
            line = line.replace('[1]','').replace('"', '').strip().split('\n')
            it = iter(line)
            for systems in it:
                sys = systems.split()[1:]
                value = next(it).split()[-1]
                if '*' not in value:
                    value = float(value)
                tra_copy.at[sys[0],sys[1]] = value
                tra_copy = tra_copy.fillna('-')
            if 'HUMAN' in tra_copy:
                tra_copy = tra_copy.drop(index='HUMAN', columns='HUMAN')

        print(tra_copy.round(5))

    f.close()

File: wilcoxon.R
#!/usr/bin/env Rscript
# This script is to computate the WMT18 way of clustering.
# Mike Zhang


# Data
files = list.files(path="wilcoxon/", pattern=".+csv", full.names=TRUE)
for (f in files) {
  mydf <- read.csv(f, sep="\t")
  cou = gsub(".+([a-z]{2})[a-z]{2}\\.csv", "\\1", f)
  trs = gsub(".+[a-z]{2}([a-z]{2})\\.csv", "\\1", f)
  df <- mydf[mydf$LANG==cou,]
  dftrs <- mydf[mydf$LANG==trs,]
  
  #averages
  
  s_avg <- aggregate( mydf$RAW.SCR ~ mydf$SYS + mydf$SID, FUN = mean )
  names(s_avg) <- c("SYS", "SID", "RAW.SCR")
  zs_avg <- aggregate( mydf$Z.SCR ~ mydf$SYS + mydf$SID, FUN = mean )
  names(zs_avg) <- c("SYS", "SID", "Z.SCR")

  s_avg_cou <- aggregate( df$RAW.SCR ~ df$SYS + df$SID, FUN = mean)
  names(s_avg_cou) <- c("SYS", "SID", "RAW.SCR")
  zs_avg_cou <- aggregate( df$Z.SCR ~ df$SYS + df$SID, FUN = mean)
  names(zs_avg_cou) <- c("SYS", "SID", "Z.SCR")

  s_avg_trs <- aggregate( dftrs$RAW.SCR ~ dftrs$SYS + dftrs$SID, FUN = mean)
  names(s_avg_trs) <- c("SYS", "SID", "RAW.SCR")
  zs_avg_trs <- aggregate( dftrs$Z.SCR ~ dftrs$SYS + dftrs$SID, FUN = mean)
  names(zs_avg_trs) <- c("SYS", "SID", "Z.SCR")

## Original Scores

s_sys <- aggregate(s_avg$RAW.SCR ~ s_avg$SYS, FUN = mean )
names(s_sys) <- c("SYS", "RAW.SCR")

s_sys_cou <- aggregate(s_avg_cou$RAW.SCR ~ s_avg_cou$SYS, FUN = mean )
names(s_sys_cou) <- c("SYS", "RAW.SCR")

s_sys_trs<- aggregate(s_avg_trs$RAW.SCR ~ s_avg_trs$SYS, FUN = mean )
names(s_sys_trs) <- c("SYS", "RAW.SCR")

# print(f)

for (sys1 in unique(zs_avg$SYS)) {
  for (sys2 in unique(zs_avg$SYS)) {
    if (identical(sys1, sys2)) {
      next
    }

# Wilcoxon rank-sum
print(paste("Comparing: ", sys1, sys2, sep=" "))
res = wilcox.test(zs_avg_trs$Z.SCR[zs_avg_trs$SYS==sys1], zs_avg_trs$Z.SCR[zs_avg_trs$SYS==sys2], conf.int=TRUE, alternative = "greater")

# 2018 metrics
if (identical("wmt18", gsub(".+(wmt[0-9]{2}).+", "\\1", f))) {
    score = abs(mean(zs_avg_trs$Z.SCR[zs_avg_trs$SYS==sys2]) - mean(zs_avg_trs$Z.SCR[zs_avg_trs$SYS=="HUMAN"])) - abs(mean(zs_avg_trs$Z.SCR[zs_avg_trs$SYS==sys1]) - mean(zs_avg_trs$Z.SCR[zs_avg_trs$SYS=="HUMAN"]))
    if (res$p.value < 0.001) {
      print(paste(round(score,2),'***',sep=""))
    } else if (res$p.value < 0.01) {
      print(paste(round(score,2),'**',sep=""))
    } else if (res$p.value < 0.05) {
      print(paste(round(score,2),'*',sep=""))
    } else {
      print(round(score,2))
    }
  } else {
  print(paste("p:", res$p.value, sep=" "))
}
    }
  }
}
print('wilcoxon')
